### PR TITLE
HPCC-19164 Component hpcc user limits set at startup

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

start/stop HPCC components many times with various edits to /etc/security/limits.conf and verified limits of processes started where properly set.

### CODE REVIEW

**Comment by mckellyln** [comment]
@Michael-Gardner please review and test on CentOS.

File: initfiles/bash/etc/init.d/hpcc_common.in [hunk 1]
```diff
@@ -468,6 +468,20 @@ startCmd() {
         fi
     fi
 
+    # apply user limits we explicitly added ...
+    myid=$(id -u 2>/dev/null)
+    if [[ "${myid}" = "0" ]] ; then
+        nofile_user=$(su - ${user} -c 'bash -c "ulimit -Hn"' 2>/dev/null)
```
Comment (mckellyln) [inline initfiles/bash/etc/init.d/hpcc_common.in]: NOTE: could this ever block, expecting input ?

File: initfiles/bash/etc/init.d/hpcc_common.in [hunk 2]
```diff
@@ -468,6 +468,20 @@ startCmd() {
...
+        rc=$?
+        if [[ $rc -ne 0 ]] ; then
+            nofile_user=$(grep 'hpcc* hard* nofile' /etc/security/limits.conf 2>/dev/null | awk '{print $4}')
```
Comment (mckellyln) [inline initfiles/bash/etc/init.d/hpcc_common.in]: NOTE: what if hpcc user limits where instead in /etc/security/limits.d/* file(s) ?

File: initfiles/bash/etc/init.d/hpcc_common.in [hunk 3]
```diff
@@ -468,6 +468,20 @@ startCmd() {
...
+        fi
+        if [[ -n "${nofile_user}" ]] ; then
+            ulimit -Hn ${nofile_user} >/dev/null 2>&1
+        fi
+    fi
+
+    # check limits are at least as high as the minimums required ...
     limits=(
```
Comment (mckellyln) [inline initfiles/bash/etc/init.d/hpcc_common.in]: NOTE: or should we just change these values to what is found in /etc/security/limits.conf or limits.d/* ?

File: initfiles/bash/etc/init.d/hpcc_common.in (hunk 2 above)
Comment (Michael-Gardner) [reply initfiles/bash/etc/init.d/hpcc_common.in]: Yeah I think we should check for the limits.d file as well,

**Comment by mckellyln** [comment]
@Michael-Gardner updated.  Once we finalize this we should extend to the other limits used (nproc etc.)

**Comment by mckellyln** [comment]
@Michael-Gardner updated.

File: initfiles/bash/etc/init.d/hpcc_common.in [hunk 4]
```diff
@@ -478,20 +507,34 @@ startCmd() {
     local i=0
     for element in "${limits[@]}"; do
         flag="-${element:4:2}"
+        name=$(echo ${element} | awk -F= '{print $1}' | cut -c 8-)
```
Comment (mckellyln) [inline initfiles/bash/etc/init.d/hpcc_common.in]: Is there a more accepted parsing way for this ?

Comment (mckellyln) [reply initfiles/bash/etc/init.d/hpcc_common.in]: This works:
```
    name=$(echo ${element%=*} | cut -c 8-)
```

Comment (mckellyln) [reply initfiles/bash/etc/init.d/hpcc_common.in]: As does this:
```
    name1=${element%=*}
    name=${name1##*_}
```

Comment (mckellyln) [reply initfiles/bash/etc/init.d/hpcc_common.in]: Updated to use this method to be more consistent with the surrounding code

**Comment by Michael-Gardner** [comment]
@mckellyln Looks like everything works.  I snagged the output of ulimit -a and -aH before and after the code you modified, and then did a diff.  Looks like its properly modifying all the limits as expected.

**Comment by mckellyln** [comment]
squashed and ready for merge

**Pull request merged**